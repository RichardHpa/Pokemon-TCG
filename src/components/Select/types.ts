export type OptionProps = {
  value?: string
  label: string
} & React.ComponentPropsWithoutRef<'option'>

export type SelectProps = {
  options: OptionProps[]
  onChange?: (val: string) => void
} & React.ComponentPropsWithoutRef<'select'>
