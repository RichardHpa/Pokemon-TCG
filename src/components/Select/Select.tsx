import { forwardRef } from 'react'

import type { SelectProps } from './types'

export const Select = forwardRef<HTMLSelectElement, SelectProps>(
  ({ options, onChange, ...props }, ref) => {
    return (
      <select
        onChange={onChange}
        ref={ref}
        className='bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500'
        {...props}
      >
        {options.map((option) => {
          const { label, value, ...optionProps } = option
          return (
            <option key={label} value={value} {...optionProps}>
              {label}
            </option>
          )
        })}
      </select>
    )
  },
)

Select.displayName = 'Select'
