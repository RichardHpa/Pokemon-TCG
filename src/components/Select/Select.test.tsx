import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { Select } from './Select'

const onChange = jest.fn()
const defaultOptions = [
  {
    label: 'test',
    value: '1',
  },
  {
    label: 'test2',
    value: '2',
  },
]

describe('Select', () => {
  test('renders Select', () => {
    const { container } = render(<Select options={defaultOptions} />)
    expect(screen.getByRole('combobox')).toBeInTheDocument()
    const options = screen.getAllByRole('option')
    expect(options).toHaveLength(2)

    expect(container).toMatchSnapshot()
  })

  test('onChange gets called when new option is selected', async () => {
    render(<Select options={defaultOptions} onChange={onChange} />)
    const select = screen.getByRole('combobox')
    expect(select).toBeInTheDocument()

    expect(onChange).not.toHaveBeenCalled()

    await userEvent.selectOptions(select, [defaultOptions[1].value])

    expect(onChange).toHaveBeenCalled()
    expect(onChange).toHaveBeenCalledWith('2')
  })
})
