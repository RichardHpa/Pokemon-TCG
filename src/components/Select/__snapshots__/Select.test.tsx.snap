// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders Select 1`] = `
<div>
  <select
    class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
  >
    <option
      value="1"
    >
      test
    </option>
    <option
      value="2"
    >
      test2
    </option>
  </select>
</div>
`;
